plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/HuaweiCloudMetricsService.java

```diff
@@ -0,0 +1,271 @@
+            Metric metric =
+                    huaweiCloudToXpanseDataModelConverter.convertShowMetricDataResponseToMetric(
+                            deployResource, showMetricDataResponse, metricInfoList);
+            metrics.add(metric);
+            cacheMetricData(metric, metrics, deployResource, metricInfoList);
+        }
+        return metrics;
+    }
+
+    private void cacheMetricData(Metric metric, List<Metric> metrics, DeployResource deployResource,
+                                 MetricInfoList metricInfoList) {
+        if (!CollectionUtils.isEmpty(metric.getMetrics())) {
+            metrics.add(metric);
+            huaweiCloudMonitorCache.set(deployResource.getResourceId(), metric);
+        } else {
+            List<Metric> metricCache =
+                    huaweiCloudMonitorCache.get(deployResource.getResourceId(),
+                            metricInfoList.getMetricName());
+            metrics.addAll(metricCache);
+        }
+    }
+
+    /**
+     * Get metrics of the @deployService.
+     *
+     * @param serviceMetricRequest The request model to query metrics.
+     * @return Returns list of metric result.
+     */
+    public List<Metric> getMetricsByService(ServiceMetricRequest serviceMetricRequest) {
+        List<Metric> metrics = new ArrayList<>();
+        List<DeployResource> deployResources = serviceMetricRequest.getDeployResources();
+        CredentialDefinition credential = serviceMetricRequest.getCredential();
+        MonitorResourceType monitorResourceType = serviceMetricRequest.getMonitorResourceType();
+        ICredential icredential = getIcredential(credential);
+        CesClient client = huaweiCloudMonitorClient.getCesClient(icredential,
+                deployResources.get(0).getProperties().get("region"));
+        Map<String, List<MetricInfoList>> deployResourceMetricInfoMap = new HashMap<>();
+        for (DeployResource deployResource : deployResources) {
+            clearExpiredMetricCache(deployResource.getResourceId());
+            List<MetricInfoList> targetMetricInfoList =
+                    getTargetMetricInfoList(deployResource, monitorResourceType, client);
+            deployResourceMetricInfoMap.put(deployResource.getResourceId(), targetMetricInfoList);
+        }
+        BatchListMetricDataRequest batchListMetricDataRequest =
+                huaweiCloudToXpanseDataModelConverter.buildBatchListMetricDataRequest(
+                        serviceMetricRequest, deployResourceMetricInfoMap);
+        BatchListMetricDataResponse batchListMetricDataResponse =
+                client.batchListMetricData(batchListMetricDataRequest);
+        metrics =
+                huaweiCloudToXpanseDataModelConverter.convertBatchListMetricResponseToMetric(
+                        batchListMetricDataResponse, deployResourceMetricInfoMap, deployResources,
+                        metrics);
+        if (!CollectionUtils.isEmpty(metrics)) {
```

Comment:
can we move just the setting cache logic here to separate private method and call it from here? otherwise the method looks too long.